File: src/selection.lua
local StructureMath = require("world/structureMath")
local CircleMenu = require("circleMenu")
local Building = require("syntheinrust").building
local vector = require("syntheinrust").vector

local Selection = {}
Selection.__index = Selection

function Selection.create(world, team)
	local self = {}
	setmetatable(self, Selection)

	self.world = world
	self.team = team

	self.build = nil
	self.sturcture = nil
	self.partIndex = nil

	self.assign = nil

	self.buildingOnStructureListeners = {}
	self.doneBuildingOnStructureListeners = {}

	return self
end

function Selection:cursorpressed(cursor, control)
	local structure = self.world:getObject(cursor.x, cursor.y)
	local part
	if structure then part = structure:findPart(cursor.x, cursor.y) end
	if structure and structure.type == "structure" and part then
		local build = self.build
		local team = structure.body:getUserData().team
		if build then
			if control.ship == "build" then
				if build.mode == 3 then
					if team == 0 or team == self.team then
						self.structure = structure
						self.part = part
						if build:setStructure(structure, part) then
							self.structure = nil
							self.part = nil
							self.build = nil
						end
						self:signalBuildingOnStructure()
					end
				end
			elseif control.ship == "destroy" then
				self.structure = nil
				self.part = nil
				self.build = nil
			end
		elseif self.assign then
			self.assign.leader = structure
			self.assign = nil
		else
			if control.ship == "build" then
				if team ~= 0 then
					local corePart = structure.corePart
					if corePart == part then
						if team == self.team and corePart.modules.drone then
							self.structure = structure
							self.part = part
						end
					end
				else
					self.build = Building.create()
					self.build:setAnnexee(structure, part)
					self.structure = structure
					self.part = part
				end
			elseif control.ship == "destroy" then
				local corePart = structure.corePart
				if team == 0 or (team == self.team and part ~= corePart) then
					structure:disconnectPart(part.location)
				end
			end
		end
	else
		if control.ship == "destroy" then
			self.structure = nil
			self.part = nil
			self.build = nil
		end
	end
end

function Selection:cursorreleased(cursor, control)
	local structure = self.structure
	local part = self.part
	if structure and part then
		local l = part.location
		local partSide, withinPart = StructureMath.getPartSide(structure, l, cursor.x, cursor.y)
		local build = self.build
		if not withinPart then
			if build then
				if structure:testEdge({ l[1], l[2], partSide }) then
					build:setSide(partSide)
					if build.mode == 5 then
						self.build = nil
						self:signalDoneBuildingOnStructure()
					end
				else
					self.build = nil
				end
			else
				local body = structure.body
				local x, y = body:getWorldPoints(l[1], l[2])
				local strength = part:getMenu()
				local newAngle = vector.angle(cursor.x - x, cursor.y - y)
				local index = CircleMenu.angleToIndex(newAngle, #strength)
				local option = self.part:runMenu(index, body)
				if option == "assign" then
					self.assign = self.part
				end
			end
			self.structure = nil
			self.partSide = nil
		else
			if not build then
				self.structure = nil
				self.partSide = nil
			end
		end
	end
end

function Selection:draw(cursor, zoom)
	local structure = self.structure
	local part = self.part
	local build = self.build
	if structure and part then
		local location = part.location
		local partX, partY = unpack(location)
		local body = structure.body
		local menuRotation -- Body angle if building else 0

		local strength, labels
		if build then
			menuRotation = body:getAngle()
			strength = {}
			local indexReverse = {1, 4, 3, 2}
			local x, y = body:getWorldPoints(partX, partY)
			local menuToCursorAngle = vector.angle(cursor.worldX - x, cursor.worldY - y)
			local partSide = CircleMenu.angleToIndex(-menuRotation + menuToCursorAngle, 4)
			local l = {partX, partY}
			for i = 1,4 do
				l[3] = indexReverse[i]
				local _, partB, connection = structure:testEdge(l)
				local connectable = not partB and connection
				local highlight = i == partSide
				local brightness = highlight and 2 or 1
				strength[i] = connectable and brightness or 0
			end
		else
			menuRotation = 0
			local x, y = body:getWorldPoints(partX, partY)
			strength, labels = part:getMenu()
			local menuToCursorAngle = vector.angle(cursor.worldX - x, cursor.worldY - y)
			local index = CircleMenu.angleToIndex(menuToCursorAngle, #strength)
			if strength[index] == 1 then
				strength[index] = 2
			end
		end
		if strength then
			local x, y = body:getWorldPoints(partX, partY)
			CircleMenu.draw(x, y, menuRotation, 1, strength, labels)
		end
	end
	if build then

		local body = build.body
		local vec = build.annexeeBaseVector
		if body and vec and build.mode > 2 then
			local l = StructureMath.addDirectionVector(vec, vec[3], .5)
			local x, y = body:getWorldPoint(l[1], l[2])
			local angle = body:getAngle()

			love.graphics.draw(
				cursor.image,
				x, y, angle,
				1/zoom, 1/zoom,
				halfCursorWidth, halfCursorWidth)
		end
	end
	local assign = self.assign
	if assign then
		local body = assign.modules.hull.fixture:getBody()
		local x, y  = body:getPosition()
		local angle = body:getAngle()

		love.graphics.draw(
			cursor.image,
			x, y, angle,
			1/zoom, 1/zoom,
			halfCursorWidth, halfCursorWidth)
	end
end

function Selection:isBuildingOnStructure()
	return self.build and self.build.structure
end

function Selection:whenBuildingOnStructure(func)
	table.insert(self.buildingOnStructureListeners, func)
end

function Selection:signalBuildingOnStructure()
	for _, func in ipairs(self.buildingOnStructureListeners) do
		func()
	end
end

function Selection:whenDoneBuildingOnStructure(func)
	table.insert(self.doneBuildingOnStructureListeners, func)
end

function Selection:signalDoneBuildingOnStructure()
	for _, func in ipairs(self.doneBuildingOnStructureListeners) do
		func()
	end
end

return Selection

File: src/selection.rs
use std::collections::HashMap;

use mlua::prelude::{LuaError, LuaFunction, LuaResult, LuaTable, LuaValue};
use mlua::{FromLua, IntoLua, Lua, UserData, UserDataFields, UserDataMethods};

use crate::building::Building;
use crate::world::shipparts::part::Location;

// TODO: Replace f64 indexes with integers
pub struct Selection {
    world: LuaTable,
    team: f64,
    building: Option<Building>,
    pub structure: Option<LuaTable>,
    structure_part: Option<LuaTable>,
    building_on_structure_listeners: Vec<LuaFunction>,
    done_building_on_structure_listeners: Vec<LuaFunction>,
}

struct Point {
    x: f64,
    y: f64,
}

pub fn new(world: LuaTable, team: f64) -> Selection {
    Selection {
        world: world,
        team: team,
        building: None,
        structure: None,
        structure_part: None,
        building_on_structure_listeners: Vec::new(),
        done_building_on_structure_listeners: Vec::new(),
    }
}

impl Selection {
    pub fn cursor_pressed(
        &mut self,
        cursor: Point,
        control: HashMap<String, String>,
    ) -> LuaResult<()> {
        let clicked_structure: Option<LuaTable> = self
            .world
            .get::<LuaFunction>("getObject")?
            .call::<Option<LuaTable>>((self.world.clone(), cursor.x, cursor.y))?
            .filter(|clicked_object_user_data| {
                clicked_object_user_data
                    .get::<String>("type")
                    .is_ok_and(|object_type| object_type == "structure")
            });
        let clicked_part: Option<LuaTable> = clicked_structure
            .clone()
            .map(|clicked_structure| clicked_structure.get::<LuaFunction>("findPart"))
            .transpose()?
            .map(|find_part| find_part.call((clicked_structure.clone(), cursor.x, cursor.y)))
            .transpose()?;

        match (clicked_structure, clicked_part) {
            (Some(clicked_structure), Some(clicked_part)) => {
                let clicked_structure_team = clicked_structure.get::<f64>("team")?;
                match &mut self.building {
                    Some(ref mut building) => {
                        match control.get("ship").map(|control| control.as_str()) {
                            Some("build") => {
                                if clicked_structure_team == 0.0
                                    || clicked_structure_team == self.team
                                {
                                    match building.set_structure(&clicked_structure, &clicked_part)
                                    {
                                        Ok(()) => {
                                            self.structure = Some(clicked_structure);
                                            self.structure_part = Some(clicked_part);
                                        }
                                        Err(err) => {
                                            // TODO: Replace eprintln to proper log message
                                            eprintln!("Debug: {}", err);
                                            self.building = None;
                                        }
                                    }
                                }
                            }
                            Some("destroy") => {
                                self.structure = None;
                                self.structure_part = None;
                                self.building = None;
                            }
                            _ => {}
                        }
                    }
                    None => match control.get("ship").map(|control| control.as_str()) {
                        Some("build") => {}
                        Some("destroy") => {
                            let core_part =
                                clicked_structure.get::<Option<LuaTable>>("corePart")?;
                            if core_part.is_none()
                                || (core_part.is_some_and(|core_part| core_part != clicked_part)
                                    && clicked_structure_team == self.team)
                            {
                                clicked_structure
                                    .get::<LuaFunction>("disconnectPart")?
                                    .call::<()>((
                                        clicked_structure,
                                        clicked_part.get::<LuaValue>("location"),
                                    ));
                            };
                        }
                        _ => {}
                    },
                }
            }
            _ => match control.get("ship").map(|control| control.as_str()) {
                Some("build") => {}
                Some("destroy") => {
                    self.structure = None;
                    self.structure_part = None;
                    self.building = None;
                }
                _ => {}
            },
        };

        Ok(())
    }

    pub fn cursor_released(
        &mut self,
        cursor: Point,
        controls: HashMap<String, String>,
    ) -> LuaResult<()> {
        Ok(())
    }
}

impl UserData for Selection {
    fn add_fields<F: UserDataFields<Self>>(fields: &mut F) {
        fields.add_field_method_get("structure", |_, this| Ok(this.structure.clone()));
    }

    fn add_methods<M: UserDataMethods<Self>>(methods: &mut M) {
        methods.add_method_mut(
            "cursorpressed",
            |_, this, (cursor, controls): (Point, HashMap<String, String>)| {
                this.cursor_pressed(cursor, controls)
            },
        );

        methods.add_method_mut(
            "cursorreleased",
            |_, this, (cursor, controls): (Point, HashMap<String, String>)| {
                this.cursor_released(cursor, controls)
            },
        );

        methods.add_method_mut("whenBuildingOnStructure", |_, this, func: LuaFunction| {
            Ok(this.building_on_structure_listeners.push(func))
        });

        methods.add_method_mut(
            "whenDoneBuildingOnStructure",
            |_, this, func: LuaFunction| Ok(this.done_building_on_structure_listeners.push(func)),
        );

        methods.add_method("draw", |_, _, ()| Ok(()));
    }
}

impl FromLua for Point {
    fn from_lua(value: LuaValue, _: &Lua) -> LuaResult<Self> {
        match value {
            LuaValue::Table(table) => Ok(Point {
                x: table.get("x")?,
                y: table.get("y")?,
            }),
            _ => Err(LuaError::FromLuaConversionError {
                from: value.type_name(),
                to: "Point".to_string(),
                message: Some("expected table".to_string()),
            }),
        }
    }
}

pub fn lua_module(lua: &Lua) -> LuaResult<LuaTable> {
    let exports = lua.create_table()?;

    exports.set(
        "create",
        lua.create_function(|lua: &Lua, (world, team): (LuaTable, f64)| {
            new(world, team).into_lua(lua)
        })?,
    )?;

    Ok(exports)
}
